File: AIClient_Python/src/world/Character.py
'''
Created on Dec 20, 2014

@author: samuel
'''
from event.DropMineEvent import DropMineEvent
from aiclient.Singleton import Singleton
from event.QueueController import QueueController
from world.Missile import Missile
from mathUtils.Vector2 import Vector2
from event.MoveCharacterEvent import MoveCharacterEvent
from event.ThrowMissileEvent import ThrowMissileEvent, Direction

class Character(object):
    queueController = Singleton(QueueController)
    
    def __init__(self, characterId, position):
        self.characterId = characterId
        self.position = Vector2(position.x, position.y)
        self.mineReady = True
        self.missile = Missile(Vector2(0,0))
        self.life = 3
    
    def move(self, position):
        self.position.x = position.x
        self.position.y = position.y
        
    def goTo(self, position):
        event1 = MoveCharacterEvent(self.characterId, position.x, position.y)
        self.queueController.outEvents.put(event1)
        
    def dropMine(self):
        if(self.mineReady):
            event = DropMineEvent(self.characterId)
            self.queueController.addOutgoingEvent(event)
            self.mineReady = False

    def hitByMine(self):
        if self.isAlive():
            self.life -= 1
        print("ouch!! - Mine")
        
    def mineHit(self):
        self.mineReady = True;
        print("Mine blow!")
        
    def shootMissile(self, direction):
        if(self.missile.isReady):
            event = ThrowMissileEvent(self.characterId, direction)
            self.queueController.addOutgoingEvent(event)
            
    def shootUp(self):
        self.shootMissile(Direction.UP)
        
    def shootDown(self):
        self.shootMissile(Direction.DOWN)
        
    def shootLeft(self):
        self.shootMissile(Direction.LEFT)
        
    def shootRight(self):
        self.shootMissile(Direction.RIGHT)
        
    def hitByMissile(self):
        if self.isAlive():
            self.life -= 1
        print("ouch!! - Missile")
        
    def missileHit(self):
        self.missile.isReady = True;
        print("Missile hit target")
        
    def isAlive(self):
        return self.life > 0
    

File: AIClient_JAVA/src/world/Character.java
/* ------------------------------------------------------------------------------
 ** _________   _________      ________    _____      _____  ___________ _________
 ** \_   ___ \ /   _____/     /  _____/   /  _  \    /     \ \_   _____//   _____/
 ** /    \  \/ \_____  \     /   \  ___  /  /_\  \  /  \ /  \ |    __)_ \_____  \ 
 ** \     \____/        \    \    \_\  \/    |    \/    Y    \|        \/        \
 **  \______  /_______  /     \______  /\____|__  /\____|__  /_______  /_______  /
 **         \/        \/             \/         \/         \/        \/        \/ 
 **
 ** Character.java
 ** A character
 **
 ** Author: Samuel-Ricardo Carriere
 ** ------------------------------------------------------------------------------*/

package world;

import java.awt.Point;

import event.DropMineEvent;
import event.MoveCharacterEvent;
import event.QueueController;
import event.ShootMissileEvent;

public class Character {
	private Point position;
	private int id;
	private int life;
	private Mine mine = new Mine();
	private Missile missile = new Missile();

	public Character(int id) {
		this.id = id;
		this.position = new Point(0, 0);
		this.life = 3;
	}

	public void move(Point position) {
		QueueController queueController = QueueController.getInstance();
		MoveCharacterEvent event = new MoveCharacterEvent(id, position);
		queueController.addOutgoingEvent(event);
	}

	public void dropMine() {
		if (mine.isReady()) {
			QueueController queueController = QueueController.getInstance();
			DropMineEvent event = new DropMineEvent(id);
			queueController.addOutgoingEvent(event);
			mine.drop();
		}
	}

	public void shootMissile(Missile.Direction direction) {
		if (missile.isReady()) {
			QueueController queueController = QueueController.getInstance();
			ShootMissileEvent event = new ShootMissileEvent(id,
					direction.ordinal());
			queueController.addOutgoingEvent(event);
			missile.shoot();
		}
	}

	public void updateInfo(Point position) {
		this.position = position;
	}

	public void hitByMine() {
		if (isAlive()) {
			--life;
			//System.out.println("Character " + id + " hit a mine");
		}
	}

	public void hitByMissile() {
		if (isAlive()) {
			--life;
			//System.out.println("Character " + id + " hit a missile");
		}
	}

	public void mineHit() {
		mine.hit();
		//System.out.println("Character " + id + " mine hit target");
	}

	public void missileHit() {
		missile.hit();
		//System.out.println("Character " + id + " missile hit target");
	}

	public Point getPosition() {
		return position;
	}

	public boolean isAlive() {
		return life > 0;
	}

	public Missile getMissile() {
		return missile;
	}
}
